include/simics/dmllib.h

```diff
@@ -2386,6 +2387,27 @@ UNUSED static void *_get_shared_memoized_outs(
     return entry;
 }
 
+UNUSED static void _independent_memoized_init_mutex(pthread_mutex_t *lock) {
+    pthread_mutexattr_t attributes;
+    int error = pthread_mutexattr_init(&attributes);
+    ASSERT(error == 0);
+    error = pthread_mutexattr_settype(&attributes, PTHREAD_MUTEX_RECURSIVE);
+    ASSERT(error == 0);
+    error = pthread_mutex_init(lock, &attributes);
+    ASSERT(error == 0);
+    error = pthread_mutexattr_destroy(&attributes);
+    ASSERT(error == 0);
+}
+
+UNUSED static void _independent_memoized_lock(
```

Comment:
Threading is hard -- please document invariants thoroughly.

Looks like this one either returns with *ran==1 and no lock taken, or with *ran!=1 and the lock taken? Would make sense to return `*ran==1`?